<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Provides helper functions for microbiome preprocessing and analysis ‚Ä¢ mbtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Provides helper functions for microbiome preprocessing and analysis">
<meta property="og:description" content="Provides helper functions for microbiome preprocessing and
    analysis.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.27.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/amplicon_sequencing.html">Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="articles/preprocessing.html">Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="articles/qa.html">Quality control of sequencing data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="inst/extdata/logo.png" width="25%" align="right"><a href="https://travis-ci.org/Gibbons-Lab/mbtools"><img src="https://travis-ci.org/Gibbons-Lab/mbtools.svg?branch=master" alt="Build Status"></a></p>
<div id="mbtools" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#mbtools" class="anchor"></a>üí© üçÉ üåé mbtools</h1></div>
<p><code>mbtools</code> is a collection of helpers that we use to analyze microbiome data. It makes it easier to run some common analyses and is pretty opinionated towards our own experiences.</p>
<div id="general-philosophy" class="section level2">
<h2 class="hasAnchor">
<a href="#general-philosophy" class="anchor"></a>General philosophy</h2>
<p><code>mbtools</code> sees analyses as a workflow consisting of several analysis steps and final outputs based on those. This is pretty similar to Qiime 2 and most of the functionality in <code>mbtools</code> is supposed to be orthogonal and not in competition to those other excellent tools.</p>
</div>
<div id="data-types" class="section level2">
<h2 class="hasAnchor">
<a href="#data-types" class="anchor"></a>Data types</h2>
<p><code>mbtools</code> mostly works on the following data types:</p>
<ol>
<li>artifacts - compound data objects returned from an analysis step</li>
<li>phyloseq object - a <a href="https://joey711.github.io/phyloseq/">phyloseq</a> object containing sequence variant abundances, taxonomy assignments and sample metadata</li>
<li>data tables - general data frame-like objects in <a href="https://r4ds.had.co.nz/tidy-data.html">tidy format</a>
</li>
</ol>
</div>
<div id="workflow-steps" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-steps" class="anchor"></a>Workflow steps</h2>
<p>For <code>mbtools</code> a workflow step is based on input data and a configuration, thus having the function signature <code><a href="https://www.rdocumentation.org/packages/stats/topics/step">step(object, config)</a></code>. Most steps can be chained with the pipe operator to yield workflows. For instance, the following is a possible workflow with <code>mbtools</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">config &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb1-2" title="2">    <span class="dt">demultiplex =</span> <span class="kw"><a href="reference/config_demultiplex.html">config_demultiplex</a></span>(<span class="dt">barcodes =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ACGTA"</span>, <span class="st">"AGCTT"</span>)),</a>
<a class="sourceLine" id="cb1-3" title="3">    <span class="dt">preprocess =</span> <span class="kw"><a href="reference/config_preprocess.html">config_preprocess</a></span>(<span class="dt">truncLen =</span> <span class="dv">200</span>),</a>
<a class="sourceLine" id="cb1-4" title="4">    <span class="dt">denoise =</span> <span class="kw"><a href="reference/config_denoise.html">config_denoise</a></span>()</a>
<a class="sourceLine" id="cb1-5" title="5">)</a>
<a class="sourceLine" id="cb1-6" title="6"></a>
<a class="sourceLine" id="cb1-7" title="7">output &lt;-<span class="st"> </span><span class="kw"><a href="reference/find_read_files.html">find_read_files</a></span>(<span class="st">"raw"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="st">          </span><span class="kw"><a href="reference/demultiplex.html">demultiplex</a></span>(config<span class="op">$</span>demultiplex) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-9" title="9"><span class="st">          </span><span class="kw"><a href="reference/quality_control.html">quality_control</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-10" title="10"><span class="st">          </span><span class="kw"><a href="reference/preprocess.html">preprocess</a></span>(config<span class="op">$</span>preprocess) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-11" title="11"><span class="st">          </span><span class="kw"><a href="reference/denoise.html">denoise</a></span>(config<span class="op">$</span>denoise)</a></code></pre></div>
<p>This clearly logs the used workflow and the configuration. The configuration can also be saved and read in many formats, for instance yaml.</p>
<p>config.yaml:</p>
<pre><code>preprocess:
  threads: yes
  out_dir: preprocessed
  trimLeft: 10.0
  truncLen: 200.0
  maxEE: 2.0
denoise:
  threads: yes
  nbases: 2.5e+08
  pool: no
  bootstrap_confidence: 0.5
  taxa_db: https://zenodo.org/record/1172783/files/silva_nr_v132_train_set.fa.gz?download=1
  species_db: https://zenodo.org/record/1172783/files/silva_species_assignment_v132.fa.gz?download=1
  hash: yes</code></pre>
<p>This can now be reused by someone else:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">config &lt;-<span class="st"> </span><span class="kw">read_yaml</span>(<span class="st">"config.yml"</span>)</a>
<a class="sourceLine" id="cb3-2" title="2"></a>
<a class="sourceLine" id="cb3-3" title="3">output &lt;-<span class="st"> </span><span class="kw"><a href="reference/find_read_files.html">find_read_files</a></span>(<span class="st">"raw"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb3-4" title="4"><span class="st">          </span><span class="kw"><a href="reference/quality_control.html">quality_control</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb3-5" title="5"><span class="st">          </span><span class="kw"><a href="reference/preprocess.html">preprocess</a></span>(config<span class="op">$</span>preprocess) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb3-6" title="6"><span class="st">          </span><span class="kw"><a href="reference/denoise.html">denoise</a></span>(config<span class="op">$</span>denoise)</a></code></pre></div>
</div>
<div id="other-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#other-functions" class="anchor"></a>Other functions</h2>
<p>All other functions are usually functions that are meant to be inside more complex code or functions that produce plots and endpoints of an analysis. Most of them act on phyloseq objects and some on tidy data tables.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/Gibbons-Lab/mbtools">https://‚Äãgithub.com/‚ÄãGibbons-Lab/‚Äãmbtools</a>
</li>
<li>Report a bug at <br><a href="https://github.com/Gibbons-Lab/mbtools/issues">https://‚Äãgithub.com/‚ÄãGibbons-Lab/‚Äãmbtools/‚Äãissues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>Apache License (== 2.0)</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Christian Diener <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>

      <footer><div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
