<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Lineage calling with SLIMM • mbtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Lineage calling with SLIMM">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.35.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_shotgun_align.html">Shotgun data - alignments</a>
    </li>
    <li>
      <a href="../articles/04_long_align.html">Long read data - alignments</a>
    </li>
    <li>
      <a href="../articles/05_lineage.html">Lineage calling with SLIMM</a>
    </li>
    <li>
      <a href="../articles/06_counting.html">Reference counting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Lineage calling with SLIMM</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Gibbons-Lab/mbtools/blob/master/vignettes/05_lineage.Rmd"><code>vignettes/05_lineage.Rmd</code></a></small>
      <div class="hidden name"><code>05_lineage.Rmd</code></div>

    </div>

    
    
<div id="lineage-calling" class="section level1">
<h1 class="hasAnchor">
<a href="#lineage-calling" class="anchor"></a>Lineage calling</h1>
<p>One thing you might want to do with metagenomics data is to identify the species present in your sample. Here we will use <a href="https://github.com/seqan/slimm">SLIMM</a> which is a binning based classifier similar to Kraken.</p>
<p>The reason we often use SLIMM is that it generates cleaned species-level coverage maps as a side effect which we use extensively.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(mbtools)</a></code></pre></div>
<pre><code>## Registered S3 methods overwritten by 'ggplot2':
##   method         from 
##   [.quosures     rlang
##   c.quosures     rlang
##   print.quosures rlang</code></pre>
<pre><code>## Registered S3 method overwritten by 'R.oo':
##   method        from       
##   throw.default R.methodsS3</code></pre>
<pre><code>## Also loading:</code></pre>
<pre><code>##   - dada2=1.12.1
##   - data.table=1.12.2
##   - ggplot2=3.1.1
##   - magrittr=1.5
##   - phyloseq=1.28.0
##   - ShortRead=1.42.0
##   - yaml=2.2.0</code></pre>
<pre><code>## Found tools:</code></pre>
<pre><code>##   - minimap2=2.15-r905
##   - slimm=0.3.4
##   - samtools=1.9</code></pre>
<pre><code>## 
## Attaching package: 'mbtools'</code></pre>
<pre><code>## The following object is masked _by_ 'package:BiocGenerics':
## 
##     normalize</code></pre>
<div id="getting-the-slimm-db" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-the-slimm-db" class="anchor"></a>Getting the SLIMM DB</h2>
<p>We will use our shotgun data again since the reference genomes are also included in the SLIMM database. For real data you will want to align against the full SLIMM database of ~15K genomes.</p>
<p><em>Download instructions will come soon :/</em></p>
<p>For now we will only use the SLIMM taxonomy mapping database which is delivered with mbtools as well.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">slimm_db &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/ABVF_SP_CMP_genomes.sldb"</span>,</a>
<a class="sourceLine" id="cb10-2" title="2">                        <span class="dt">package =</span> <span class="st">"mbtools"</span>)</a></code></pre></div>
<p>which will save the db in the folder <code>refs</code>.</p>
</div>
<div id="aligning-shotgun-reads" class="section level2">
<h2 class="hasAnchor">
<a href="#aligning-shotgun-reads" class="anchor"></a>Aligning shotgun reads</h2>
<p>One advantage of SLIMM is that it is relatively agnostic to the alignment program. So we can use our normal alignment workflow.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">fi &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/shotgun"</span>, <span class="dt">package =</span> <span class="st">"mbtools"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="st">      </span><span class="kw"><a href="../reference/find_read_files.html">find_read_files</a></span>()</a>
<a class="sourceLine" id="cb11-3" title="3">ref &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/genomes/zymo_mock.fna.gz"</span>,</a>
<a class="sourceLine" id="cb11-4" title="4">                   <span class="dt">package =</span> <span class="st">"mbtools"</span>)</a>
<a class="sourceLine" id="cb11-5" title="5"></a>
<a class="sourceLine" id="cb11-6" title="6">alns &lt;-<span class="st"> </span><span class="kw"><a href="../reference/align_short_reads.html">align_short_reads</a></span>(</a>
<a class="sourceLine" id="cb11-7" title="7">    fi,</a>
<a class="sourceLine" id="cb11-8" title="8">    <span class="dt">threads =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb11-9" title="9">    <span class="dt">reference =</span> ref,</a>
<a class="sourceLine" id="cb11-10" title="10">    <span class="dt">use_existing =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## INFO [2019-05-25 22:20:38] Aligning 3 samples on 3 threads. Keeping up to 100 secondary alignments.
## INFO [2019-05-25 22:20:40] Finished aligning even1.
## INFO [2019-05-25 22:20:42] Finished aligning even2.
## INFO [2019-05-25 22:20:44] Finished aligning even3.</code></pre>
</div>
<div id="using-slimm" class="section level2">
<h2 class="hasAnchor">
<a href="#using-slimm" class="anchor"></a>Using SLIMM</h2>
<p>SLIMM is again wrapped by a workflow so we can use it passing in an alignment artifact and a configuration.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">config &lt;-<span class="st"> </span><span class="kw"><a href="../reference/config_slimm.html">config_slimm</a></span>(</a>
<a class="sourceLine" id="cb13-2" title="2">    <span class="dt">threads =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb13-3" title="3">    <span class="dt">database =</span> slimm_db</a>
<a class="sourceLine" id="cb13-4" title="4">)</a>
<a class="sourceLine" id="cb13-5" title="5"></a>
<a class="sourceLine" id="cb13-6" title="6">config</a></code></pre></div>
<pre><code>## $reports
## NULL
## 
## $bin_width
## [1] 1000
## 
## $relative_cutoff
## [1] 0
## 
## $database
## [1] "/home/cdiener/R/x86_64-pc-linux-gnu-library/3.6/mbtools/extdata/ABVF_SP_CMP_genomes.sldb"
## 
## $threads
## [1] 3
## 
## $rank
## [1] "species"
## 
## attr(,"class")
## [1] "config"</code></pre>
<p>With this we can perform the SLIMM workflow.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">sl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/slimm.html">slimm</a></span>(alns, config)</a></code></pre></div>
<pre><code>## INFO [2019-05-25 22:20:44] running SLIMM on 3 alignments with database /home/cdiener/R/x86_64-pc-linux-gnu-library/3.6/mbtools/extdata/ABVF_SP_CMP_genomes.sldb.
## INFO [2019-05-25 22:20:45] Parsing abundances on rank `species`.
## INFO [2019-05-25 22:20:45] Estimating read lengths from a sample of 100 reads per alignment.
## INFO [2019-05-25 22:20:45] Estimated median read length is 100, range is [100, 100].
## INFO [2019-05-25 22:20:45] Parsing coverage profiles with bin width of 1000bp.</code></pre>
<p>The artifact now contains the species level lineage calls…</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">sl<span class="op">$</span>abundance</a></code></pre></div>
<pre><code>##       rank taxa_id relative reads    id              kingdom
## 1: species      0*        1 12598 even1 unknown_superkingdom
## 2: species      0*        1 12127 even2 unknown_superkingdom
## 3: species      0*        1 12346 even3 unknown_superkingdom
##            phylum         class         order         family         genus
## 1: unknown_phylum unknown_class unknown_order unknown_family unknown_genus
## 2: unknown_phylum unknown_class unknown_order unknown_family unknown_genus
## 3: unknown_phylum unknown_class unknown_order unknown_family unknown_genus
##            species
## 1: unknown_species
## 2: unknown_species
## 3: unknown_species</code></pre>
<p>Where we see no successful calls since our coverage is way to low for the example data.</p>
<p>However we can investigate which strains the data aligned to.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">sl<span class="op">$</span>coverage[, <span class="kw">uniqueN</span>(strain)] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>()</a></code></pre></div>
<pre><code>## [1] 10</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">sl<span class="op">$</span>coverage</a></code></pre></div>
<pre><code>##     bin_width           reads  genbank
##  1:      1000 0,0,0,0,0,0,... BK006935
##  2:      1000 0,0,0,0,0,0,... AE016830
##  3:      1000 0,0,0,1,0,0,... AE017341
##  4:      1000 2,2,0,0,0,0,... AL009126
##  5:      1000 0,0,0,0,0,0,... AE006468
##  6:      1000 0,2,0,0,0,0,... BA000007
##  7:      1000 0,0,0,0,0,0,... AP008937
##  8:      1000 2,0,2,0,0,0,... AL591824
##  9:      1000 0,0,0,0,0,0,... CP000253
## 10:      1000 0,2,0,0,0,0,... AE004091
## 11:      1000 0,0,0,0,0,0,... BK006935
## 12:      1000 0,0,0,0,0,2,... AE016830
## 13:      1000 0,0,0,0,0,0,... AE017341
## 14:      1000 0,0,0,0,0,0,... AL009126
## 15:      1000 0,0,0,0,4,0,... AE006468
## 16:      1000 0,0,0,0,0,2,... BA000007
## 17:      1000 0,0,0,0,0,0,... AP008937
## 18:      1000 0,0,0,0,0,0,... AL591824
## 19:      1000 2,0,0,0,0,3,... CP000253
## 20:      1000 0,0,2,0,2,0,... AE004091
## 21:      1000 0,0,0,0,0,0,... BK006935
## 22:      1000 0,0,0,0,0,0,... AE016830
## 23:      1000 0,2,0,0,0,0,... AE017341
## 24:      1000 0,0,0,0,0,0,... AL009126
## 25:      1000 0,0,0,2,0,0,... AE006468
## 26:      1000 0,0,0,0,0,0,... BA000007
## 27:      1000 0,2,0,0,0,0,... AP008937
## 28:      1000 0,0,0,0,0,0,... AL591824
## 29:      1000 0,0,0,0,2,0,... CP000253
## 30:      1000 0,0,0,0,0,0,... AE004091
##     bin_width           reads  genbank
##                                                               strain
##  1:                                   Saccharomyces cerevisiae S288C
##  2:                                       Enterococcus faecalis V583
##  3:                    Cryptococcus neoformans var. neoformans JEC21
##  4:                       Bacillus subtilis subsp. subtilis str. 168
##  5: Salmonella enterica subsp. enterica serovar Typhimurium str. LT2
##  6:                              Escherichia coli O157:H7 str. Sakai
##  7:                                 Lactobacillus fermentum IFO 3956
##  8:                                     Listeria monocytogenes EGD-e
##  9:                    Staphylococcus aureus subsp. aureus NCTC 8325
## 10:                                      Pseudomonas aeruginosa PAO1
## 11:                                   Saccharomyces cerevisiae S288C
## 12:                                       Enterococcus faecalis V583
## 13:                    Cryptococcus neoformans var. neoformans JEC21
## 14:                       Bacillus subtilis subsp. subtilis str. 168
## 15: Salmonella enterica subsp. enterica serovar Typhimurium str. LT2
## 16:                              Escherichia coli O157:H7 str. Sakai
## 17:                                 Lactobacillus fermentum IFO 3956
## 18:                                     Listeria monocytogenes EGD-e
## 19:                    Staphylococcus aureus subsp. aureus NCTC 8325
## 20:                                      Pseudomonas aeruginosa PAO1
## 21:                                   Saccharomyces cerevisiae S288C
## 22:                                       Enterococcus faecalis V583
## 23:                    Cryptococcus neoformans var. neoformans JEC21
## 24:                       Bacillus subtilis subsp. subtilis str. 168
## 25: Salmonella enterica subsp. enterica serovar Typhimurium str. LT2
## 26:                              Escherichia coli O157:H7 str. Sakai
## 27:                                 Lactobacillus fermentum IFO 3956
## 28:                                     Listeria monocytogenes EGD-e
## 29:                    Staphylococcus aureus subsp. aureus NCTC 8325
## 30:                                      Pseudomonas aeruginosa PAO1
##                                                               strain
##                      species          genus             family
##  1: Saccharomyces cerevisiae  Saccharomyces Saccharomycetaceae
##  2:    Enterococcus faecalis   Enterococcus    Enterococcaceae
##  3:  Cryptococcus neoformans   Cryptococcus    Cryptococcaceae
##  4:        Bacillus subtilis       Bacillus        Bacillaceae
##  5:      Salmonella enterica     Salmonella Enterobacteriaceae
##  6:         Escherichia coli    Escherichia Enterobacteriaceae
##  7:  Lactobacillus fermentum  Lactobacillus   Lactobacillaceae
##  8:   Listeria monocytogenes       Listeria       Listeriaceae
##  9:    Staphylococcus aureus Staphylococcus  Staphylococcaceae
## 10:   Pseudomonas aeruginosa    Pseudomonas   Pseudomonadaceae
## 11: Saccharomyces cerevisiae  Saccharomyces Saccharomycetaceae
## 12:    Enterococcus faecalis   Enterococcus    Enterococcaceae
## 13:  Cryptococcus neoformans   Cryptococcus    Cryptococcaceae
## 14:        Bacillus subtilis       Bacillus        Bacillaceae
## 15:      Salmonella enterica     Salmonella Enterobacteriaceae
## 16:         Escherichia coli    Escherichia Enterobacteriaceae
## 17:  Lactobacillus fermentum  Lactobacillus   Lactobacillaceae
## 18:   Listeria monocytogenes       Listeria       Listeriaceae
## 19:    Staphylococcus aureus Staphylococcus  Staphylococcaceae
## 20:   Pseudomonas aeruginosa    Pseudomonas   Pseudomonadaceae
## 21: Saccharomyces cerevisiae  Saccharomyces Saccharomycetaceae
## 22:    Enterococcus faecalis   Enterococcus    Enterococcaceae
## 23:  Cryptococcus neoformans   Cryptococcus    Cryptococcaceae
## 24:        Bacillus subtilis       Bacillus        Bacillaceae
## 25:      Salmonella enterica     Salmonella Enterobacteriaceae
## 26:         Escherichia coli    Escherichia Enterobacteriaceae
## 27:  Lactobacillus fermentum  Lactobacillus   Lactobacillaceae
## 28:   Listeria monocytogenes       Listeria       Listeriaceae
## 29:    Staphylococcus aureus Staphylococcus  Staphylococcaceae
## 30:   Pseudomonas aeruginosa    Pseudomonas   Pseudomonadaceae
##                      species          genus             family
##                 order               class         phylum   kingdom    id
##  1: Saccharomycetales     Saccharomycetes     Ascomycota Eukaryota even1
##  2:   Lactobacillales             Bacilli     Firmicutes  Bacteria even1
##  3:       Tremellales     Tremellomycetes  Basidiomycota Eukaryota even1
##  4:        Bacillales             Bacilli     Firmicutes  Bacteria even1
##  5:  Enterobacterales Gammaproteobacteria Proteobacteria  Bacteria even1
##  6:  Enterobacterales Gammaproteobacteria Proteobacteria  Bacteria even1
##  7:   Lactobacillales             Bacilli     Firmicutes  Bacteria even1
##  8:        Bacillales             Bacilli     Firmicutes  Bacteria even1
##  9:        Bacillales             Bacilli     Firmicutes  Bacteria even1
## 10:   Pseudomonadales Gammaproteobacteria Proteobacteria  Bacteria even1
## 11: Saccharomycetales     Saccharomycetes     Ascomycota Eukaryota even2
## 12:   Lactobacillales             Bacilli     Firmicutes  Bacteria even2
## 13:       Tremellales     Tremellomycetes  Basidiomycota Eukaryota even2
## 14:        Bacillales             Bacilli     Firmicutes  Bacteria even2
## 15:  Enterobacterales Gammaproteobacteria Proteobacteria  Bacteria even2
## 16:  Enterobacterales Gammaproteobacteria Proteobacteria  Bacteria even2
## 17:   Lactobacillales             Bacilli     Firmicutes  Bacteria even2
## 18:        Bacillales             Bacilli     Firmicutes  Bacteria even2
## 19:        Bacillales             Bacilli     Firmicutes  Bacteria even2
## 20:   Pseudomonadales Gammaproteobacteria Proteobacteria  Bacteria even2
## 21: Saccharomycetales     Saccharomycetes     Ascomycota Eukaryota even3
## 22:   Lactobacillales             Bacilli     Firmicutes  Bacteria even3
## 23:       Tremellales     Tremellomycetes  Basidiomycota Eukaryota even3
## 24:        Bacillales             Bacilli     Firmicutes  Bacteria even3
## 25:  Enterobacterales Gammaproteobacteria Proteobacteria  Bacteria even3
## 26:  Enterobacterales Gammaproteobacteria Proteobacteria  Bacteria even3
## 27:   Lactobacillales             Bacilli     Firmicutes  Bacteria even3
## 28:        Bacillales             Bacilli     Firmicutes  Bacteria even3
## 29:        Bacillales             Bacilli     Firmicutes  Bacteria even3
## 30:   Pseudomonadales Gammaproteobacteria Proteobacteria  Bacteria even3
##                 order               class         phylum   kingdom    id
##       length read_length
##  1: 12073000          NA
##  2:  3219000          NA
##  3: 19053000          NA
##  4:  4216000          NA
##  5:  4858000          NA
##  6:  5499000          NA
##  7:  2099000          NA
##  8:  2945000          NA
##  9:  2822000          NA
## 10:  6265000          NA
## 11: 12073000          NA
## 12:  3219000          NA
## 13: 19053000          NA
## 14:  4216000          NA
## 15:  4858000          NA
## 16:  5499000          NA
## 17:  2099000          NA
## 18:  2945000          NA
## 19:  2822000          NA
## 20:  6265000          NA
## 21: 12073000          NA
## 22:  3219000          NA
## 23: 19053000          NA
## 24:  4216000          NA
## 25:  4858000          NA
## 26:  5499000          NA
## 27:  2099000          NA
## 28:  2945000          NA
## 29:  2822000          NA
## 30:  6265000          NA
##       length read_length</code></pre>
<p>Where we see we actually identified the correct 10 strains. Just the coverage is really low.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/hist">hist</a></span>(sl<span class="op">$</span>coverage<span class="op">$</span>reads[<span class="dv">1</span>][[<span class="dv">1</span>]])</a></code></pre></div>
<p><img src="05_lineage_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Where we see that the majority of that genome has not been mapped.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#lineage-calling">Lineage calling</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-the-slimm-db">Getting the SLIMM DB</a></li>
      <li><a href="#aligning-shotgun-reads">Aligning shotgun reads</a></li>
      <li><a href="#using-slimm">Using SLIMM</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
