<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>6 - Reference counting • mbtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="6 - Reference counting">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.39.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_shotgun_align.html">3 - Shotgun data - alignments</a>
    </li>
    <li>
      <a href="../articles/04_long_align.html">4 - Long read data - alignments</a>
    </li>
    <li>
      <a href="../articles/05_lineage.html">5 - Lineage calling with SLIMM</a>
    </li>
    <li>
      <a href="../articles/06_counting.html">6 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/07_workflow_management.html">7 - Workflow management</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>6 - Reference counting</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Gibbons-Lab/mbtools/blob/master/vignettes/06_counting.Rmd"><code>vignettes/06_counting.Rmd</code></a></small>
      <div class="hidden name"><code>06_counting.Rmd</code></div>

    </div>

    
    
<div id="reference-counting" class="section level1">
<h1 class="hasAnchor">
<a href="#reference-counting" class="anchor"></a>Reference counting</h1>
<p>One basic anaylisis we want to perform is to count the abundance of some references in our data. Here reference is a broad term and may refer to transcripts, genomes, or amplicon sequences contained in a reference database.</p>
<p>For this <code>mbtools</code> implements a general purpose expectation maximization (EM) counter. It is implemented in C++ to perform fast and usually reading the alignment file takes longer than the actual analysis.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(mbtools)</span></code></pre></div>
<pre><code>## Also loading:</code></pre>
<pre><code>##   - dada2=1.12.1
##   - data.table=1.12.2
##   - ggplot2=3.2.1
##   - magrittr=1.5
##   - phyloseq=1.28.0
##   - ShortRead=1.42.0
##   - yaml=2.2.0</code></pre>
<pre><code>## Found tools:</code></pre>
<pre><code>##   - minimap2=2.17-r941
##   - slimm=0.3.4
##   - samtools=1.9</code></pre>
<pre><code>## 
## Attaching package: 'mbtools'</code></pre>
<pre><code>## The following object is masked _by_ 'package:BiocGenerics':
## 
##     normalize</code></pre>
<div id="alignment" class="section level2">
<h2 class="hasAnchor">
<a href="#alignment" class="anchor"></a>Alignment</h2>
<p>Let’s start by aligning our short read data to a reference database of 10 microbial genomes.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>fi &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/shotgun"</span>, <span class="dt">package =</span> <span class="st">"mbtools"</span>) <span class="op">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="st">      </span><span class="kw"><a href="../reference/find_read_files.html">find_read_files</a></span>()</span>
<span id="cb8-3"><a href="#cb8-3"></a>ref &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/genomes/zymo_mock.fna.gz"</span>,</span>
<span id="cb8-4"><a href="#cb8-4"></a>                   <span class="dt">package =</span> <span class="st">"mbtools"</span>)</span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a>alns &lt;-<span class="st"> </span><span class="kw"><a href="../reference/align_short_reads.html">align_short_reads</a></span>(</span>
<span id="cb8-7"><a href="#cb8-7"></a>    fi,</span>
<span id="cb8-8"><a href="#cb8-8"></a>    <span class="dt">threads =</span> <span class="dv">3</span>,</span>
<span id="cb8-9"><a href="#cb8-9"></a>    <span class="dt">reference =</span> ref,</span>
<span id="cb8-10"><a href="#cb8-10"></a>    <span class="dt">use_existing =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>## INFO [2019-08-28 13:00:14] Aligning 3 samples on 3 threads. Keeping up to 100 secondary alignments.
## INFO [2019-08-28 13:00:16] Finished aligning even1.
## INFO [2019-08-28 13:00:18] Finished aligning even2.
## INFO [2019-08-28 13:00:20] Finished aligning even3.</code></pre>
</div>
<div id="em-counting" class="section level2">
<h2 class="hasAnchor">
<a href="#em-counting" class="anchor"></a>EM counting</h2>
<p>Counting is yet again a workflow and requires an alignment artifact and a configuration.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>config &lt;-<span class="st"> </span><span class="kw"><a href="../reference/config_count.html">config_count</a></span>(</span>
<span id="cb10-2"><a href="#cb10-2"></a>    <span class="dt">reference =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/genomes/zymo_mock.fna.gz"</span>,</span>
<span id="cb10-3"><a href="#cb10-3"></a>                            <span class="dt">package =</span> <span class="st">"mbtools"</span>),</span>
<span id="cb10-4"><a href="#cb10-4"></a>    <span class="dt">threads =</span> <span class="dv">1</span>,</span>
<span id="cb10-5"><a href="#cb10-5"></a>    <span class="dt">weights =</span> <span class="ot">TRUE</span></span>
<span id="cb10-6"><a href="#cb10-6"></a>)</span>
<span id="cb10-7"><a href="#cb10-7"></a></span>
<span id="cb10-8"><a href="#cb10-8"></a>config</span></code></pre></div>
<pre><code>## $reference
## [1] "/Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/genomes/zymo_mock.fna.gz"
## 
## $threads
## [1] 1
## 
## $method
## [1] "em"
## 
## $maxit
## [1] 10000
## 
## $cutoff
## [1] 0.01
## 
## $tpm
## [1] FALSE
## 
## $weights
## [1] TRUE
## 
## attr(,"class")
## [1] "config"</code></pre>
<p>And we can proceed to counting:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>cn &lt;-<span class="st"> </span><span class="kw"><a href="../reference/count_references.html">count_references</a></span>(alns, config)</span></code></pre></div>
<pre><code>## INFO [2019-08-28 13:00:20] Getting reference lengths from /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/genomes/zymo_mock.fna.gz...
## INFO [2019-08-28 13:00:21] Normalized IDs. Starting counting...
## INFO [2019-08-28 13:00:21] [alignments/even1.bam] Read 16941 alignments.
## INFO [2019-08-28 13:00:21] [alignments/even1.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-08-28 13:00:21] [alignments/even1.bam] Used 20 EM iterations on 16 equivalence classes. Last max. abs. change was 0. Database concordance is 0.16%.
## INFO [2019-08-28 13:00:22] [alignments/even2.bam] Read 16804 alignments.
## INFO [2019-08-28 13:00:22] [alignments/even2.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-08-28 13:00:22] [alignments/even2.bam] Used 20 EM iterations on 12 equivalence classes. Last max. abs. change was 1.1e-13. Database concordance is 0.17%.
## INFO [2019-08-28 13:00:22] [alignments/even3.bam] Read 17518 alignments.
## INFO [2019-08-28 13:00:22] [alignments/even3.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-08-28 13:00:22] [alignments/even3.bam] Used 20 EM iterations on 20 equivalence classes. Last max. abs. change was 5.7e-14. Database concordance is 0.16%.</code></pre>
<p>This creates a count artifact which contains the counts.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>cn<span class="op">$</span>counts</span></code></pre></div>
<pre><code>##      reference    counts effective_length sample
##  1: AE004091.2 633.56663          6264306  even1
##  2: AE006468.2 669.75553          4857352  even1
##  3: AE016830.1 660.15548          3217933  even1
##  4: AE017341.1 603.14526         19052864  even1
##  5: AL009126.3 627.85029          4215508  even1
##  6: AL591824.1 628.19900          2944430  even1
##  7: AP008937.1 463.50018          2098587  even1
##  8: BA000007.3 702.28268          5498480  even1
##  9: BK006935.2 583.40725         12072428  even1
## 10: CP000253.1 716.85064          2821263  even1
## 11:       &lt;NA&gt;  10.28706               NA  even1
## 12: AE004091.2 637.96080          6264306  even2
## 13: AE006468.2 546.44977          4857352  even2
## 14: AE016830.1 684.74418          3217933  even2
## 15: AE017341.1 588.45327         19052864  even2
## 16: AL009126.3 587.75863          4215508  even2
## 17: AL591824.1 510.03447          2944430  even2
## 18: AP008937.1 670.60417          2098587  even2
## 19: BA000007.3 630.70269          5498480  even2
## 20: BK006935.2 608.11390         12072428  even2
## 21: CP000253.1 590.07601          2821263  even2
## 22:       &lt;NA&gt;  10.10210               NA  even2
## 23: AE004091.2 616.13757          6264306  even3
## 24: AE006468.2 652.73269          4857352  even3
## 25: AE016830.1 599.92729          3217933  even3
## 26: AE017341.1 638.01323         19052864  even3
## 27: AL009126.3 556.38136          4215508  even3
## 28: AL591824.1 719.03635          2944430  even3
## 29: AP008937.1 731.04321          2098587  even3
## 30: BA000007.3 553.32322          5498480  even3
## 31: BK006935.2 610.70098         12072428  even3
## 32: CP000253.1 485.69589          2821263  even3
## 33:       &lt;NA&gt;  10.00820               NA  even3
##      reference    counts effective_length sample</code></pre>
<p>Note that there is a NA reference. This is the number of reads likely coming from a sequence not contained in the reference.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>cn<span class="op">$</span>counts[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(reference)]</span></code></pre></div>
<pre><code>##    reference   counts effective_length sample
## 1:      &lt;NA&gt; 10.28706               NA  even1
## 2:      &lt;NA&gt; 10.10210               NA  even2
## 3:      &lt;NA&gt; 10.00820               NA  even3</code></pre>
<p>In each sample we have 10 microbes with the same abundance.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">ggplot</span>(cn<span class="op">$</span>counts, <span class="kw">aes</span>(<span class="dt">x=</span>counts, <span class="dt">y=</span>reference)) <span class="op">+</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">xlim</a></span>(<span class="dv">0</span>, <span class="dv">1000</span>) <span class="op">+</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>sample) <span class="op">+</span><span class="st"> </span><span class="kw">theme_minimal</span>()</span></code></pre></div>
<p><img src="06_counting_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Looks ok, but there is a lot of noise since we have very low depth.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="kw">ggplot</span>(cn<span class="op">$</span>counts, <span class="kw">aes</span>(<span class="dt">x =</span> counts)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>(<span class="dt">bins =</span> <span class="dv">8</span>)</span></code></pre></div>
<p><img src="06_counting_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>We could also compare that with a naive counting method that just uses the best alignment score. In our case that will perform equally since we have little multi-mapping here.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>config<span class="op">$</span>method &lt;-<span class="st"> "naive"</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>cn2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/count_references.html">count_references</a></span>(alns, config)</span></code></pre></div>
<pre><code>## INFO [2019-08-28 13:00:23] Getting reference lengths from /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/genomes/zymo_mock.fna.gz...
## INFO [2019-08-28 13:00:23] Normalized IDs. Starting counting...
## INFO [2019-08-28 13:00:23] [alignments/even1.bam] Read 16941 alignments.
## INFO [2019-08-28 13:00:23] [alignments/even1.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-08-28 13:00:23] [alignments/even2.bam] Read 16804 alignments.
## INFO [2019-08-28 13:00:23] [alignments/even2.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].
## INFO [2019-08-28 13:00:24] [alignments/even3.bam] Read 17518 alignments.
## INFO [2019-08-28 13:00:24] [alignments/even3.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>counts &lt;-<span class="st"> </span>cn<span class="op">$</span>counts[cn2<span class="op">$</span>counts, on =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"sample"</span>, <span class="st">"reference"</span>, <span class="st">"effective_length"</span>)]</span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="kw">ggplot</span>(counts, <span class="kw">aes</span>(<span class="dt">x =</span> counts, <span class="dt">y =</span> i.counts)) <span class="op">+</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">"lm"</span>) <span class="op">+</span></span>
<span id="cb22-5"><a href="#cb22-5"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">"em"</span>, <span class="dt">y =</span> <span class="st">"naive"</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>()</span></code></pre></div>
<p><img src="06_counting_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>As we see they both give the same results for that simple case.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#reference-counting">Reference counting</a><ul class="nav nav-pills nav-stacked">
<li><a href="#alignment">Alignment</a></li>
      <li><a href="#em-counting">EM counting</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
