<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Runs a full DADA2 workflow. — denoise • mbtools</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Runs a full DADA2 workflow. — denoise" />

<meta property="og:description" content="This performs the following steps:
Learning the error rate for each run separately
Inferring sequence variants for each run and sample
Merging of feature tables across runs
Consensus chimera removal
Taxa assignment with Naive Bayes
Species assignment by exact alignment
Diagnostic plots of the error rates
You will usually want to preprocess the read files first with
preprocess.
Depending on your config the sequences might be represented by
an MD5 hash. In this case the taxa table has an additional `sequence` column
containing the real sequences." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.26.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/amplicon_sequencing.html">Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/preprocessing.html">Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/qa.html">Quality control of sequencing data</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Runs a full DADA2 workflow.</h1>
    <small class="dont-index">Source: <a href='https://github.com/Gibbons-Lab/mbtools/blob/master/R/workflow_dada2.R'><code>R/workflow_dada2.R</code></a></small>
    <div class="hidden name"><code>denoise.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This performs the following steps:</p><ol>
<li><p>Learning the error rate for each run separately</p></li>
<li><p>Inferring sequence variants for each run and sample</p></li>
<li><p>Merging of feature tables across runs</p></li>
<li><p>Consensus chimera removal</p></li>
<li><p>Taxa assignment with Naive Bayes</p></li>
<li><p>Species assignment by exact alignment</p></li>
<li><p>Diagnostic plots of the error rates</p></li>
</ol><p>You will usually want to preprocess the read files first with
<code><a href='preprocess.html'>preprocess</a></code>.
Depending on your config the sequences might be represented by
an MD5 hash. In this case the taxa table has an additional `sequence` column
containing the real sequences.</p>
    
    </div>

    <pre class="usage"><span class='fu'>denoise</span>(<span class='no'>object</span>, <span class='no'>config</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>An experiment data table as returned by
<code><a href='find_read_files.html'>find_read_files</a></code> or a worflow object.</p></td>
    </tr>
    <tr>
      <th>config</th>
      <td><p>A configuration file as returned by
<code><a href='config_denoise.html'>config_denoise</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list containing the workflow results:</p><dl class='dl-horizontal'>
  <dt>feature_table</dt><dd><p>Matrix of sequence variant abundances. Samples are
   and sequences are columns.</p></dd>
  <dt>taxonomy</dt><dd><p>Matrix of taxonomy assignments. Rows are sequences and
   columns are taxonomy ranks.</p></dd>
  <dt>errors</dt><dd><p>The error profiles estimated by DADA2. One for each run and
   read direction (forward/reverse).</p></dd>
  <dt>error_plots</dt><dd><p>Plotted error profiles. One for each run and
   read direction (forward/reverse).</p></dd>
  <dt>passed_reads</dt><dd><p>How many reads were kept in each step. Rows are
    samples and columns are workflow steps.</p></dd>
  <dt>classified</dt><dd><p>The proportion of sequence variants that could be
       where a specific taxa rank could be classified.</p></dd>
</dl>

    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

